### PR TITLE
pycubedmini library updates

### PR DESCRIPTION
**Pull request by yashikabatra15**
Changes:
- Initializing sun sensors and h bridges in separate try-except statements. This allows us to pinpoint what exactly is failing
- Added burnwire as an entry in the hardware dictionary (init function in Satellite class)
- General documentation and style (local variable names, spacing, comments). I didn’t change global variable names because I’m not sure where else these are accessed yet. I think it would be best to refactor once we have all the code ready, and this isn’t a priority.
- Reinit function:
    - Line 216: dev.lower() -> dev.strip().lower(). This change strips the string dev of whitespace and then makes it all lowercase, which helps us avoid small errors (like extra spaces in commands)
    - Line 220: self.radio2… -> self.radio… This is because we only have one radio on Pycubedmini (I suspect that it says radio2 because it wasn’t updated from PyCubed)
- Burn function: Current version used variable names that are never initialized in pycubedmini (but do exist in PyCubed). Cleaned up this function to support the current variables and hardware. 

Necessary Updates:
- General: Add the camera board as an entry in the hardware dictionary (init function)
- storage_stats function: Return the storage statistics for the Nicla board as well, as it has its own file system

Possible Updates:
- acceleration, magnetic, gyro, temperature functions. 2 options:
    1. return the reading when this function is called (this is currently the case)
    2. store readings and return readings with matching timestamps
- Burn function: We currently have to pass a string or an array of strings as burn_num: is there anything stopping us from passing an integer?

### CODE REVIEW

File: state_machine/plugs/flight/lib/pycubed.py [hunk 1]
```diff
@@ -78,7 +77,8 @@ def __init__(self):
             'GPS':    False,
             'WDT':    False,
             'Sun':    False,
-            'Coils':  False
+            'Coils':  False,
+            'BurnWire': False
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:81]: This is a bit pedantic but generally speaking to reduce the future number of lines changes in commit we want to put a comma even on the last entry to the dictionary:
ie ```
a = {
    "1": True,
    "2": False,
}
If the previous author had stuck to this convention adding the `'BurnWire': False,` would have only been one insertion (rather than 1 deletion and two insertions)

File: state_machine/plugs/flight/lib/pycubed.py [hunk 2]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+            sun_sensors.append(sun_yp)
+        except Exception as e:
+            print('[ERROR][Sun Sensor +Y]', e)
+            
+        try:
             sun_zp = adafruit_tsl2561.TSL2561(self.i2c1, address=0x39)  # +Z
+            sun_sensors.append(sun_zp)
+        except Exception as e:
+            print('[ERROR][Sun Sensor +Z]', e)
+
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:181]: Generally I like this, however there is a lot of code duplication. From my understanding we have 6 code blocks like these: 
```python
try:
    sun_yn = adafruit_tsl2561.TSL2561(self.i2c2, address=0x29)  # -Y
except Exception as e:
    print('[ERROR][Sun Sensor -Y]', e)
```

each of which only varies by the address, I think this should be done with a loop

File: state_machine/plugs/flight/lib/pycubed.py [hunk 3]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+            coils.append(drv_x)
+        except Exception as e:
+            print('[ERROR][H-Bridge U6]', e)
+            
+        try:
             drv_y = drv8830.DRV8830(self.i2c3, 0x60)  # U8
+            coils.append(drv_y)
+        except Exception as e:
+            print('[ERROR][H-Bridge U8]', e)
+
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:212]: Similar situation here with code duplication as with the sun sensors

File: state_machine/plugs/flight/lib/pycubed.py [hunk 4]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+            self.burnwire1 = pwmio.PWMOut(microcontroller.pin.PA15, frequency=1000, duty_cycle=0)
+            burnwires.append(self.burnwire1)
         except Exception as e:
-            print('[ERROR][H-Bridges]', e)
+            print('[ERROR][Burn Wire IC1]', e)
+        
+        try:
+            # needed to change pinout from BURN2 to PA18, as BURN2 did not support PWMOut
+            self.burnwire2 = pwmio.PWMOut(microcontroller.pin.PA18, frequency=1000, duty_cycle=0)
+            burnwires.append(self.burnwire2)
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:240]: Similar code duplication with burn wires.

File: state_machine/plugs/flight/lib/pycubed.py [hunk 5]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+        
+        if burnwire_count >= 1:
+            self.hardware['BurnWire'] = True
+        
 
+    # reinit: reinitialize radio, sd, or IMU based upon the contents of string dev
     def reinit(self, dev):
-        dev = dev.lower()
+        # dev is a whitespace-stripped string of all lowercase letters
+        dev = dev.strip().lower()
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:255]: This seems strange to have `reinit("               IMU")`, `reinit("imu")`, `reinit("iMu   ")`, and `reinit('IMU')` have the same behavior.

File: state_machine/plugs/flight/lib/pycubed.py [hunk 6]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+            burnwire_count += 1
+        
+        if burnwire_count >= 1:
+            self.hardware['BurnWire'] = True
+        
 
+    # reinit: reinitialize radio, sd, or IMU based upon the contents of string dev
     def reinit(self, dev):
-        dev = dev.lower()
+        # dev is a whitespace-stripped string of all lowercase letters
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:254]: From my understanding `strip` doesn't remove all whitespace, but only the whitespace from the beginning and end.

File: state_machine/plugs/flight/lib/pycubed.py [hunk 7]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
             self._sd.__init__(self.spi, self._sdcs, baudrate=1000000)
         elif dev == 'imu':
             self.IMU.__init__(self.i2c1)
         else:
-            print('Invalid Device? ->', dev)
+            print('Invalid Device? ->',dev)
+
 
     @property
+    # return the accelerometer reading from the IMU
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:270]: For these please use docstrings, rather than comments as this enables better IDE integration.

File: state_machine/plugs/flight/lib/pycubed.py [hunk 8]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
     def acceleration(self):
         return self.IMU.accel
 
+
     @property
+    # return the magnetometer reading from the IMU
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:276]: Should be docstring

File: state_machine/plugs/flight/lib/pycubed.py [hunk 9]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
     def magnetic(self):
         return self.IMU.mag
 
+
     @property
+    # return the gyroscope reading from the IMU
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:282]: Should be docstring

File: state_machine/plugs/flight/lib/pycubed.py [hunk 10]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
     def gyro(self):
         return self.IMU.gyro
 
+
     @property
+    # return the thermometer reading from the IMU
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:288]: should be docstring

File: state_machine/plugs/flight/lib/pycubed.py [hunk 11]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
     def temperature(self):
-        return self.IMU.temperature  # Celsius
+        return self.IMU.temperature # Celsius
+
 
     @property
+    # return the temperature reading from the CPU
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:294]: docstring

File: state_machine/plugs/flight/lib/pycubed.py [hunk 12]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
     def temperature_cpu(self):
-        return microcontroller.cpu.temperature  # Celsius
+        return microcontroller.cpu.temperature # Celsius
+
 
     @property
+    # return the current RBG settings of the neopixel object
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:300]: docstring

File: state_machine/plugs/flight/lib/pycubed.py [hunk 13]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
     def RGB(self):
         return self.neopixel[0]
 
+    
     @RGB.setter
-    def RGB(self, value):
+    # set an RGB value to the neopixel object
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:306]: docstring

File: state_machine/plugs/flight/lib/pycubed.py [hunk 14]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
         if self.hardware['Neopixel']:
             try:
                 self.neopixel[0] = value
             except Exception as e:
-                print('[WARNING]', e)
+                print('[WARNING]',e)
+
 
     @property
+    # return the battery voltage
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:316]: docstring

File: state_machine/plugs/flight/lib/pycubed.py [hunk 15]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+        vbat=0
+
         for _ in range(50):
-            vbat += self._vbatt.value * 3.3 / 65536
-        _voltage = (vbat / 50) * (100 + 100) / 100  # 100k/100k voltage divider
-        return _voltage  # volts
+            # 65536 = 2^16, number of increments we can have to voltage
+            vbat+=self._vbatt.value * 3.3 / 65536
+        
+        # 316/110 voltage divider
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:325]: These are the incorrect resistor values, stick to the updated values from here:
```python
_voltage = (vbat / 50) * (100 + 100) / 100  # 100k/100k voltage divider
```

File: state_machine/plugs/flight/lib/pycubed.py [hunk 16]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+        voltage = (vbat/50)*(316+110)/110
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:326]: Same resistor value issue but not in the comment

File: state_machine/plugs/flight/lib/pycubed.py [hunk 17]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
-        '''
-        substring something like '/data/DATA_'
-        directory is created on the SD!
-        int padded with zeros will be appended to the last found file
-        '''
+
+    # create a new file on the SD card
+    # substring example: '/data/DATA_'
+    # int padded with zeroes will be appended to the last found file
+    def new_file(self,substring, binary=False):
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:403]: docstring

File: state_machine/plugs/flight/lib/pycubed.py [hunk 18]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+                # create new filepath in the sd directory, using the given folder and file names
+                self.filename = '/sd' + folder + filen + "{:03}".format(n + 1) + ".txt"
+            
+            # create new file with open, write timestamp and status
+            with open(self.filename, "a") as f:
+                f.write('# Created: {:.0f}\r\n# Status: {}\r\n'.format(time.monotonic(), self.status))
+
+            # print a confirmation that this new file was created
             print('New self.filename:', self.filename)
-            return self.filename
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:323]: why did you change it to return a boolean rather than the file name?

File: state_machine/plugs/flight/lib/pycubed.py [hunk 19]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+            with open(self.logfile, "a") as l:
+                l.write('# Created: {:.0f}\r\n# Status: {}\r\n'.format(time.monotonic(), self.status))
+
+            # print a confirmation message that a new logfile was created
             print('New log file:', self.logfile)
 
-    def print_file(self, filedir=None):
-        if filedir is None:
+
+    # print a file given its directory; file directory is by default None
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:500]: doc string

File: state_machine/plugs/flight/lib/pycubed.py [hunk 20]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
             for line in file:
                 print(line.strip())
 
-    def send_file(self, c_size, send_buffer, filename):
-        num_packets = int(stat(filename)[6] / c_size)
-        with open(filename, "rb") as f:
-            for i in range(num_packets + 1):
-                f.seek(i * c_size)
+
+    # send a file given character size, buffer size, and the filename
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:514]: docstring

File: state_machine/plugs/flight/lib/pycubed.py [hunk 21]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
-                yield bytes([i, 0x45, num_packets])
 
+                # return bytes; yield keyword returns without destroying states of local vars
+                yield bytes([i,0x45,num_packets])
+
+
+    # save the passed dataset to the passed savefile
+    # dataset should be a set of lists; each line is a list: save(([line1],[line2]))
+    # to save a string, make it an item in a list: save(['This is my string'])
+    # by default, savefile is not passed
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:534]: Doc string

**Review by thetazero — CHANGES_REQUESTED** [review]
Really good initial work here, really like most of the changes. I've commented on most of the things that seem off, it's mostly just comments instead of docstring and a few other nuanced things. I didn't comment on this but the file does need to be linted.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 2 above)
Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:181]: Yashika and I discussed this. I think we keep it as it is for now and we'll modify it once we actually know how we'll want to be accessing the sun sensors, as that is going to change while we develop the attitude determination code.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 3 above)
Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:212]: Same as above - keep it as is for now, clean it up when we change the attitude control code.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 5 above)
Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:255]: What is the purpose of the reinit function?

Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:255]: (As in where does it get called from). I'm not really sure what the use case is where having this string cleansing would be helpful.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 2 above)
Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:181]: In reality, we should have each of these initializations broken out into individual functions
- `_init_sun_sensors()`
-  `_init_coil_drivers()`
- etc.
that all get called from `__init__`.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 15 above)
Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:325]: good catch

File: state_machine/plugs/flight/lib/pycubed.py [hunk 22]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+        # volts
+        return voltage
+
 
     @property
+    # report battery voltage as % full
     def fuel_gauge(self):
-        # report battery voltage as % full
-        return 100 * self.battery_voltage / 8.4
+        return 100*self.battery_voltage/8.4
```
[the remark this answers is not in the document]
Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:335]: Nope it's definitely 4.2. I believe the cubesat-scale `pycubed.py` has two lithium ions in series which is where the 8.4 comes from.

**Review by jbwillis — CHANGES_REQUESTED** [review]

File: state_machine/plugs/flight/lib/pycubed.py (hunk 5 above)
Comment (thetazero) [reply state_machine/plugs/flight/lib/pycubed.py:255]: The only reference I remember to it is when transitioning out of low power mode, so realistically string cleansing should never be helpful from my understanding.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 2 above)
Comment (yashikabatra15) [reply state_machine/plugs/flight/lib/pycubed.py:181]: I like the idea of breaking down the initializations into individual functions - I can update the file to reflect this if it's a change we want now. Generally, I think it's a good idea to get all the major changes to this file done before we refactor, because pycubedmini.py is accessed in a LOT of different places.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 5 above)
Comment (yashikabatra15) [reply state_machine/plugs/flight/lib/pycubed.py:255]: Line 255 originally used dev.lower(), so I added dev.strip() to avoid more string issues (which is what I presume .lower() was called for to begin with). I weirdly enough couldn't find a place where reinit was called where passing a string would be relevant, but I kept it as is / added .strip() in the case we later choose to use reinit in a way where this would matter.

File: state_machine/plugs/flight/lib/pycubed.py [hunk 23]
```diff
@@ -147,289 +147,467 @@ def __init__(self):
...
+    def burn(self, burn_num, dutycycle = 0, freq = 1000, duration = 1):
+        # BURN1 = -Z,BURN2 = extra burnwire pin, dutycycle ~0.13%
         dtycycl = int((dutycycle / 100) * (0xFFFF))
+
+        # print configuration information
         print('----- BURN WIRE CONFIGURATION -----')
-        print(f'\tFrequency of: {freq}Hz')
-        print(f'\tDuty cycle of: {100 * dtycycl / 0xFFFF}% (int:{dtycycl})')
-        print(f'\tDuration of {duration}sec')
+        print('\tFrequency of: {}Hz\n\tDuty cycle of: {}% (int:{})\n\tDuration of {}sec'.format(freq,(100 * dtycycl / 0xFFFF), dtycycl, duration))
```
[the remark this answers is not in the document]
Comment (yashikabatra15) [reply state_machine/plugs/flight/lib/pycubed.py:586]: I do not remember editing it this way -- that's my bad, I'll fix it.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 2 above)
Comment (thetazero) [reply state_machine/plugs/flight/lib/pycubed.py:181]: I think we should do this now

File: state_machine/plugs/flight/lib/pycubed.py (hunk 5 above)
Comment (thetazero) [reply state_machine/plugs/flight/lib/pycubed.py:264]: I really don't see a world in which we accidentally pass `"                                 iMu         "` instead of `"IMU"`. Can you please give me an example in which it would result in expected behavior?

**Review by thetazero — CHANGES_REQUESTED** [review]
Looks much better, still would like you to finish linting the file with flake8. I've already fixed most of the issues in #21, so it would be unfortunate to undo all that progress. I think it's fine to leave a few things unfixed as they require changing functionality like #17.

File: state_machine/plugs/flight/lib/pycubed.py [hunk 24]
```diff
@@ -13,12 +11,14 @@
...
 import busio
 import time
+# import json
 import digitalio
 import analogio
 import storage
 import sys
-import pulseio
 import neopixel
+# import pulseio
```
Comment (jbwillis) [inline state_machine/plugs/flight/lib/pycubed.py:20]: If we aren't using the library, just delete the line rather than comment it out

File: state_machine/plugs/flight/lib/pycubed.py [hunk 25]
```diff
@@ -13,12 +11,14 @@
 import microcontroller
 import busio
 import time
+# import json
```
Comment (jbwillis) [inline state_machine/plugs/flight/lib/pycubed.py:14]: If we aren't using the library, just delete the line rather than comment it out

File: state_machine/plugs/flight/lib/pycubed.py [hunk 26]
```diff
@@ -41,34 +41,125 @@
...
+    c_boot = multiBitFlag(register=_BOOTCNT, lowest_bit=0, num_bits=8)
+    c_state_err = multiBitFlag(register=_RSTERRS, lowest_bit=4, num_bits=4)
+    c_vbus_rst = multiBitFlag(register=_RSTERRS, lowest_bit=0, num_bits=4)
+    c_deploy = multiBitFlag(register=_DCOUNT, lowest_bit=0, num_bits=8)
+    c_downlink = multiBitFlag(register=_DWNLINK, lowest_bit=0, num_bits=8)
 
+    # change to 433?
     UHF_FREQ = 433.0
+    
+    def __init_sun_sensors(self, i2c1, i2c2):
```
Comment (jbwillis) [inline state_machine/plugs/flight/lib/pycubed.py:67]: I'd say let's put the `__init_xxx` functions after the `__init__` function.

Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:67]: Also, since `self` has access to `i2c1` and `i2c2`, why do we need to pass them into this function separately?

**Review by jbwillis — CHANGES_REQUESTED** [review]

File: state_machine/plugs/flight/lib/pycubed.py [hunk 27]
```diff
@@ -220,216 +317,339 @@ def RGB(self, value):
...
     def new_file(self, substring, binary=False):
-        '''
-        substring something like '/data/DATA_'
-        directory is created on the SD!
-        int padded with zeros will be appended to the last found file
-        '''
+        """
+        create a new file on the SD card
+        substring example: '/data/DATA_'
+        int padded with zeroes will be appended to the last found file
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:417]: This is a little confusing to me

File: state_machine/plugs/flight/lib/pycubed.py [hunk 28]
```diff
@@ -220,216 +317,339 @@ def RGB(self, value):
...
+        # if size of current open logfile > 100MB, create new log file
+        if stat(self.logfile)[6] > 1E8:
             self.new_Log()
+
+        # open the current logfile and write message msg with a timestamp
         if self.hardware['SDcard']:
             with open(self.logfile, "a+") as file:
                 file.write('{:.1f},{}\r\n'.format(time.monotonic(), msg))
 
     def new_Log(self):
```
Comment (thetazero) [inline state_machine/plugs/flight/lib/pycubed.py:502]: We may want to consider renaming this, as it breaks the snake_case convention held throughout all other methods.

**Review by thetazero — CHANGES_REQUESTED** [review]
Still some minor linting errors that can be fixed (at this point just letting vscode auto format should do it). Other than that looks pretty much set.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 28 above)
Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:502]: Yeah, I agree, `new_log` would be better

File: state_machine/plugs/flight/lib/pycubed.py (hunk 27 above)
Comment (jbwillis) [reply state_machine/plugs/flight/lib/pycubed.py:417]: Since this is mostly being carried over from the previous file, let's leave it alone. I agree that this function is a bit odd, but I think we should resolve that later.

**Review by jbwillis — APPROVED** [review]

**Comment by jbwillis** [comment]
@yashikabatra15 please run this on the mainboard before merging - we definitely want some level of testing for this.

File: state_machine/plugs/flight/lib/pycubed.py (hunk 28 above)
Comment (yashikabatra15) [reply state_machine/plugs/flight/lib/pycubed.py]: Fixed!

**Comment by yashikabatra15** [comment]
I've reviewed / implemented all changes and linted the file! Please let me know if there's anything else y'all see before we merge.

**Review by thetazero — APPROVED** [review]
